### PR TITLE
Introduce improved SROA pass

### PR DESCRIPTION
**Pull request by Keno**
This is a rebased and fixed version of the improved SROA pass from #26778.
There's a decent piece of new infrastructure wrapped up in this: The ability
to insert new nodes during compaction. This is a bit tricky because it requires
tracking which version of the statements buffer a given SSAValue belongs to.
At the moment this is done mostly manually, but I'm hoping to clean that up
in the future. The idea of the new SROA pass is fairly straightforward:
Given a use of an interesting value, it traces through all phi nodes, finding
all leaves, applies whatever transformation to those leaves and then re-inserts
a phi nest corresponding to the phi nest of the original value.

### CODE REVIEW

**Comment by Keno** [comment]
I apologize in advance for some of the messiness in this PR, but since it's on the critical path for 0.7alpha, my plan is to merge this as soon as acceptable by CI (and nanosoldier), so we can merge #25261 and tag the alpha. There's a whole bunch of cleanup to be done after.

**Comment by Keno** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

File: base/compiler/ssair/verify.jl [hunk 1]
```diff
@@ -28,6 +28,10 @@ function check_op(ir::IRCode, domtree::DomTree, @nospecialize(op), use_bb::Int,
                 error()
             end
         end
+    elseif isa(op, Union{OldSSAValue, NewSSAValue})
+        #@Base.show ir
+        @verify_error "Left over SSA marker"
+        error()
```
Comment (vtjnash) [inline base/compiler/ssair/verify.jl]: need to fix these `error()` calls before release

File: base/compiler/ssair/ir.jl [hunk 2]
```diff
@@ -455,52 +472,147 @@ function getindex(compact::IncrementalCompact, idx::Int)
...
+            elseif isa(val, NewSSAValue)
+                needs_late_fixup = true
             end
         end
     end
+    needs_late_fixup
+end
+
+function resort_pending!(compact)
+    sort!(compact.pending_perm, DEFAULT_STABLE, Order.By(x->compact.pending_nodes[x].pos))
```
Comment (vtjnash) [inline base/compiler/ssair/ir.jl]: is this supposed to be returning `compact.pending_perm`?

File: base/compiler/ssair/ir.jl [hunk 3]
```diff
@@ -455,52 +472,147 @@ function getindex(compact::IncrementalCompact, idx::Int)
...
+end
+
+function insert_node!(compact::IncrementalCompact, before, @nospecialize(typ), @nospecialize(val), attach_after::Bool=false)
```
Comment (vtjnash) [inline base/compiler/ssair/ir.jl]: type signature, or use dispatch

File: base/compiler/ssair/ir.jl [hunk 4]
```diff
@@ -455,52 +472,147 @@ function getindex(compact::IncrementalCompact, idx::Int)
...
+        before_entry = compact.new_new_nodes[before.id]
+        push!(compact.new_new_nodes, NewNode(before_entry.pos, attach_after, typ, val, before_entry.line))
+        return NewSSAValue(length(compact.new_new_nodes))
+    else
+        error("Unsupported")
+    end
 end
 
-function insert_node_here!(compact::IncrementalCompact, @nospecialize(val), @nospecialize(typ), ltable_idx::Int)
+function insert_node_here!(compact::IncrementalCompact, @nospecialize(val), @nospecialize(typ), ltable_idx::Int, reverse_affinity=false)
```
Comment (vtjnash) [inline base/compiler/ssair/ir.jl]: type signature

File: base/compiler/ssair/ir.jl [hunk 5]
```diff
@@ -455,52 +472,147 @@ function getindex(compact::IncrementalCompact, idx::Int)
...
+            @assert compact.used_ssas[val.id] >= 1
+            compact.used_ssas[val.id] -= 1
+        end
+    end
+    compact.result[idx.id] = v
+    # Add count for new use
+    if count_added_node!(compact, v)
+        push!(compact.late_fixup, idx.id)
+    end
 end
```
Comment (vtjnash) [inline base/compiler/ssair/ir.jl]: `setindex!` should return first argument

File: base/compiler/ssair/ir.jl [hunk 6]
```diff
@@ -455,52 +472,147 @@ function getindex(compact::IncrementalCompact, idx::Int)
...
 
 function setindex!(compact::IncrementalCompact, @nospecialize(v), idx::Int)
```
Comment (vtjnash) [inline base/compiler/ssair/ir.jl]: `setindex!` should return first argument

File: base/compiler/ssair/ir.jl [hunk 7]
```diff
@@ -509,10 +621,14 @@ end
 
 function getindex(view::TypesView, idx)
```
Comment (vtjnash) [inline base/compiler/ssair/ir.jl]: type-signature

File: base/compiler/ssair/ir.jl [hunk 8]
```diff
@@ -528,57 +644,114 @@ function done(compact::IncrementalCompact, (idx, _a)::Tuple{Int, Int})
     return idx > length(compact.ir.stmts) && (compact.new_nodes_idx > length(compact.perm))
 end
 
+function getindex(view::TypesView, idx::NewSSAValue)
+    @assert isa(view.ir, IncrementalCompact)
+    compact = view.ir
```
Comment (vtjnash) [inline base/compiler/ssair/ir.jl]: use type-assert here for better performance and better error messages

File: base/compiler/ssair/ir.jl [hunk 9]
```diff
@@ -651,14 +824,21 @@ function next(compact::IncrementalCompact, (idx, active_bb)::Tuple{Int, Int})
         compact.new_nodes_idx += 1
         new_node_entry = compact.ir.new_nodes[new_idx]
         new_idx += length(compact.ir.stmts)
-        return process_newnode!(compact, new_idx, new_node_entry, idx, active_bb)
+        return process_newnode!(compact, new_idx, new_node_entry, idx, active_bb, true)
+    elseif !isempty(compact.pending_perm) &&
+        (entry = compact.pending_nodes[compact.pending_perm[1]];
+         entry.attach_after ? entry.pos == idx - 1 : entry.pos == idx)
```
Comment (vtjnash) [inline base/compiler/ssair/ir.jl]: can we make the control flow here less convoluted?

File: base/compiler/ssair/ir.jl [hunk 10]
```diff
@@ -673,22 +853,29 @@ function next(compact::IncrementalCompact, (idx, active_bb)::Tuple{Int, Int})
     return Pair{Int, Any}(old_result_idx, compact.result[old_result_idx]), (compact.idx, active_bb)
 end
 
-function maybe_erase_unused!(extra_worklist, compact, idx)
-    effect_free = stmt_effect_free(compact.result[idx], compact, compact.ir.mod)
+function maybe_erase_unused!(extra_worklist, compact, idx, callback = x->nothing)
```
Comment (vtjnash) [inline base/compiler/ssair/ir.jl]: type-signature

File: base/compiler/ssair/verify.jl (hunk 1 above)
Comment (Keno) [reply base/compiler/ssair/verify.jl]: This is in the verifier.

Comment (vtjnash) [reply base/compiler/ssair/verify.jl]: yes, that's why I said before release, not before merge

**Comment by Keno** [comment]
Looks fine. I'll merge this and address the review in a follow up PR.

**Comment by vtjnash** [comment]
What's the PR number of the followup? This PR re-introduced many of the performance issues common to the rest of this new code (failing to annotate types in argument signatures, using `enumerate`, etc.). It's hard to improve compiler performance it we can't even do the simple stuff. (eventually we also need to do the harder work too, like not using IdDict, but I understand that will require more rework).

**Comment by Keno** [comment]
Still on my todo list. There were a number of correctness issues to address first.

**Pull request merged**